[MRG+1] Add Davies-Bouldin index

logc opened the pull request:
Add another unsupervised quality metric for clustering results, the Davies-Bouldin Index.

#### Reference Issues/PRs
closes #7942

#### What does this implement/fix? Explain your changes.
This implements an unsupervised quality metric for clustering results, the [Davies-Bouldin Index](https://en.wikipedia.org/wiki/Davies–Bouldin_index), based on an already existing PR that was stalled. The differences between this commit and the changes proposed there are minimal. In particular, the tests are copied verbatim, to ensure that this implementation does still conform to what was expected.

#### Any other comments?
I noticed while working on a toy problem that there are not many alternatives in `sklearn` for unsupervised metrics of clustering quality. In particular, I missed Davies-Bouldin from working with other ML packages (Weka, IIRC).

Looking through ` sklearn`, I found the above mentioned PR, noticed the author @tomron does not seem to answer anymore, and decided to push for this change to get accepted by doing a similar proposal. I fixed all remaining open comments.

If there is a better way to get the DB index into `sklearn`, please tell me. If there are other comments that can still be improved in this implementation, I will do my best to correct them, too.

logc committed:
Add Davies-Bouldin index

Add another unsupervised quality metric for clustering results, the
Davies-Bouldin Index.

jnothman commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -256,3 +257,57 @@ def calinski_harabaz_score(X, labels):
+    ----------
+    .. [1] `Davies, David L.; Bouldin, Donald W. (1979).
+       "A Cluster Separation Measure". IEEE Transactions on
+       Pattern Analysis and Machine Intelligence. PAMI-1 (2): 224-227`_
```
Please add an Examples Sexton

logc replied:
I'm sorry, I can see there are sections like this in other parts of the doc, but I don't know how to generate the example contents (?)

jnothman replied:
It should just be a couple of lines showing how you would use this function in a simple case.

logc replied:
Isn't the doctest example in lines 1625-1637 doing that?

jnothman replied:
Perhaps. I believe it belongs more here, in the API documentation, than in the narrative user guide

jnothman commented on doc/modules/clustering.rst:
```diff
@@ -1591,6 +1591,90 @@ Drawbacks
+cluster analysis.
+
+  >>> from sklearn.cluster import KMeans
+  >>> from sklearn.metrics import davis_bouldin_index
```
Typo here

logc replied:
Fixed

jnothman reviewed (commented):
Thanks for this. Also, flake8 is failing

jnothman commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+                         rng.rand(10, 2), np.arange(10))
+
+    # Assert the value is 0. when all samples are equals
+    assert_equal(0., davies_bouldin_index(np.ones((10, 2)),
```
These days we would prefer a bare assert, rather than assert_equal

logc replied:
Fixed

jnothman commented:
Feel like wrapping up #8135 for us also so we don't need to add tests here?

glemaitre commented:
#10828 is taking over #8135 and could be almost merge with we don't consider to add the two tests which I mentioned in the summary.

logc committed:
Code review changes

- Add Davies-Bouldin to clustering test_common
- Fix flake8 issue
- Use bare `assert` in test
- Fix typo

jnothman commented:
@logc, I don't think you've pushed a second commit.

jnothman commented:
Please merge in the latest master, where we have just added common tests for clustering metrics. Please add this to the common tests in sklearn/metrics/cluster/tests/test_common.py

logc commented:
@jnothman sorry, commented before pushing the second commit. The tests run really long, locally (!)

jnothman commented:
> The tests run really long, locally (!)

Do you mean the full test suite? The clustering metrics tests?

logc commented:
@jnothman yes, the full test suite. Currently, I am running it with `make` from the project root, before creating a commit. If I only have changes to the `test_unsupervised.py` file, then I only run that. If there is a better strategy, please tell me.

jnothman commented:
If I were you, I'd run `pytest sklearn/metrics/cluster` and let Travis do the rest.

logc commented:
@jnothman added `davies_bouldin_index` to `test_common`. In order to pass `test_format_invariance`, I had to change the `dtype` of the centroids variable from adapting to `X` to hardcoded as float. This was one of the open questions in the previous PR. If you do not do that assumption, then the index is not the same for an array of ints and the same array cast as float.

logc commented:
@jnothman can I have another review here? :smile:

Also, @tguillemot had comments on the previously existing PR; maybe I could have a review here, too?

Thanks!

jnothman commented:
is the difference between int and float X substantial?

jnothman commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+         [[0, 4], [1, 3]] * 5 + [[3, 1], [4, 0]] * 5)
+    labels = [0] * 10 + [1] * 10 + [2] * 10 + [3] * 10
+    assert_almost_equal(davies_bouldin_index(X, labels),
+                        2*np.sqrt(0.5)/3)
```
Please include spaces around * and /

logc replied:
Done

jnothman commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+    X = ([[0, 0], [2, 2], [3, 3], [5, 5]])
+    labels = [0, 0, 1, 2]
+    assert_almost_equal(davies_bouldin_index(X, labels),
+                        (5./4)/3)
```
Spaces around /, please

logc replied:
Done

jnothman reviewed (approved):
LGTM, thanks

logc committed:
Add spaces around operators

logc commented:
My ping to @tguillemot was not very successful 😄 ; maybe @glemaitre would help here in having a second opinion? Thanks to any and all reviewers.

glemaitre commented:
I put it in my list of revision :) You should receive my review soon

glemaitre reviewed (changes requested):
Sorry for the delay.

Couple of comments and this is missing a what's new entry.

lzfelix commented:
Can you please provide supporting evidence that Davies-Bouldin index in bounded on the interval [0, 1]? The original paper by Davies and Boulding `[1]` only explicitly states that this measure is grater or equal than zero. On the other hand, the MATLAB documentation shows a [scenario where the index is larger than zero](https://www.mathworks.com/help/stats/clustering.evaluation.daviesbouldinevaluation-class.html) under K-means clustering:

```
NumObservations: 600
InspectedK: [1 2 3 4 5 6]
CriterionValues: [NaN 0.4663 0.4454 0.8316 1.0444 0.9236]
OptimalK: 3
```

**References:** `[1] Davies, David L., and Donald W. Bouldin. "A cluster separation measure." IEEE transactions on pattern analysis and machine intelligence 2 (1979): 224-227.`

glemaitre commented:
@lzfelix Reading the references, it seems that there is nothing mentioning an upper bound equal to 1.
I am not sure exactly how to make the formal proof.

My intuition to find an upper bound to 1 would be in the below configuration:
![circle_cluster](https://user-images.githubusercontent.com/7454015/39093003-91c8e360-4618-11e8-850f-6032f26797a2.jpg)

But honestly, this is just a draw and I would appreciate with somebody could have a formal proof, either way.

@jnothman @GaelVaroquaux @lesteve @agramfort

lzfelix commented:
@glemaitre it seems that `DBI > 1` if there's some degree of overlap between two clusters, at least for the bidimensional case. It's harder to figure out what happens on higher-dimensional spaces without a better analysis. Still, it is possible to obtain `DBI ~ 1.19` using the code below:

```python
X, y = sklearn.datasets.make_blobs(n_samples=400 ,centers=1, center_box=(0, 1))
y_hat = sklearn.cluster.KMeans(n_clusters=2).fit_predict(X)

plt.scatter(*X.T, c=y_hat, alpha=0.4)
print('Davies Bouldin index: {:4.4}'.format(davies_bouldin_index(X, y_hat)))
```

which produces:
![clustering_output](https://image.ibb.co/e8Axrx/clustering.png)

glemaitre commented:
Actually this is not really an overlap but more something like ellipses-like shape. In this case, it will be higher than 1. However, I am not sure what would be the upper bound formally. 

logc committed:
Second code review changes

- Fix incorrectly parsed documentation block
- Fix references indentation
- Refactor test assertions

lzfelix commented:
@glemaitre you are right. Maybe the upper bound helps on deciding if the generated clusters are not well defined, in the opposite sense that `DBI ~ 0` corresponds to a good partitioning of the data? Still, it's just speculation...

Anyways, I just wanted to try to contribute on the docs of this PR to avoid later confusion :)

glemaitre commented:
Thanks this is useful review. We don't like to put erroneous doc :)                                                                                                                                                                                   ‎

logc commented on sklearn/metrics/cluster/unsupervised.py:
```diff
@@ -256,3 +257,57 @@ def calinski_harabaz_score(X, labels):
+       "A Cluster Separation Measure". IEEE Transactions on
+       Pattern Analysis and Machine Intelligence. PAMI-1 (2): 224-227`_
+    """
+    X, labels = check_X_y(X, labels)
```
This seems a bit out of scope for this PR. Also, there are small differences in each metric that makes the refactor non-trivial. I could take it up in a later PR if you do not mind.

logc commented on doc/modules/clustering.rst:
```diff
@@ -1591,6 +1591,90 @@ Drawbacks
+   DB = \frac{1}{k} \sum{i=1}^k \max_{i \neq j} R_{ij}
+
+
+  >>> from sklearn import datasets
```
Done!

logc commented on doc/modules/clustering.rst:
```diff
@@ -1591,6 +1591,90 @@ Drawbacks
+
+.. topic:: References
+
+* Davies, David L.; Bouldin, Donald W. (1979).
```
Done! My RST skills are rusty, I only write Markdown these days ...

logc commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+    rng = np.random.RandomState(seed=0)
+
+    # Assert message when there is only one label
+    assert_raise_message(ValueError, "Number of labels is",
```
Done!

logc commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+                         rng.rand(10, 2), np.zeros(10))
+
+    # Assert message when all point are in different clusters
+    assert_raise_message(ValueError, "Number of labels is",
```
Done!

logc commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+                         rng.rand(10, 2), np.arange(10))
+
+    # Assert the value is 0. when all samples are equals
+    assert 0. == davies_bouldin_index(np.ones((10, 2)),
```
Done!

logc commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+                                      [0] * 5 + [1] * 5)
+
+    # Assert the value is 0. when all the mean cluster are equal
+    assert 0. == davies_bouldin_index([[-1, -1], [1, 1]] * 10,
```
Done!

logc commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+    X = ([[0, 0], [1, 1]] * 5 + [[3, 3], [4, 4]] * 5 +
+         [[0, 4], [1, 3]] * 5 + [[3, 1], [4, 0]] * 5)
+    labels = [0] * 10 + [1] * 10 + [2] * 10 + [3] * 10
+    assert_almost_equal(davies_bouldin_index(X, labels),
```
Done! Also, refactored other usages of `assert_almost_equal` to `pytest.approx` in this module, for consistency.

logc commented on sklearn/metrics/cluster/tests/test_unsupervised.py:
```diff
@@ -193,3 +194,38 @@ def test_calinski_harabaz_score():
+    # General case - cluster have one sample
+    X = ([[0, 0], [2, 2], [3, 3], [5, 5]])
+    labels = [0, 0, 1, 2]
+    assert_almost_equal(davies_bouldin_index(X, labels),
```
Done!

logc commented:
@lzfelix Thanks for pointing out that inconsistency! I am having a close look at the original reference, to see where this "upper bound is 1" idea is rooted ... I am not sure if it is the documentation, or the implementation which is wrong.

By the way, if I understand correctly `make_blobs`, in your example you are forcing the data to be scattered around a *single* center, and then fit a KMeans model with `n=2`. One could argue that this is a fundamental shortcoming of the algorithm -- it tries to partition the dataset always in as many labels as given. In fact, DBI is "warning", in the RuntimeWarning, that the clustering is artificial because it results in centroid distances that are 0 ... ?

logc commented:
@glemaitre How do you correctly create a "What's new" entry for this change?

lzfelix commented:
@logc the argument that DBI is warning that the clusters are malformed are rooted on the assumption that DBI should be smaller than one. On the other hand, so far we have not proven this bound nor are sure if it has a meaning, so these conclusions are not being built on solid assumptions.

Your observations are correct about `make_blobs`. I created the example this way because it was easy to understand and reproduce. On the other hand I was able to produce the same behaviour on a dataset with high-dimensional data: while `DBI > 1`, the V-measure is larger than 0.5. Still, since the data is high dimensional (and dense) it's hard to wonder what's going on there. 😞

glemaitre commented:
‎In the section new feature of the v.0.20 in the doc/whatsnew folder. You can check the format there.

logc commented:
@lzfelix you are absolutely right! In the original paper, the authors provide an example where a clustering results in an index value of 2.0. They state that this "reflects a particularly inappropriate partition", but the value is correct & expected. I will update the doc to reflect this.

logc commented:
I removed the reference to bounds in the DBI value. Now the doc only says that

> A lower index value is indicative of a better clustering partition

lzfelix commented:
@logc great! Maybe "index values closer to zero indicative of a better clustering partition"? This way you ensure to convey the message that there is a lower bound only.

jnothman commented:
Or just say "Zero is the minimum score and a greater score is better.".

On 23 April 2018 at 06:04, Luiz Felix <notifications@github.com> wrote:

> @logc <https://github.com/logc> great! Maybe "index values closer to zero
> indicative of a better clustering partition"? This way you ensure to convey
> the message that there is a lower bound only.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10827#issuecomment-383408425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6yz3ucR3nDxE56-o6I87k_mu_W_tks5trOJZgaJpZM4Su-1P>
> .
>

lzfelix commented:
Actually, the smaller the better.

jnothman commented:
of course. forgetting the context. just trying to simplify the message

logc committed:
Documentation changes

logc commented:
@glemaitre I am sure this PR is on your list, but can let me ask: is there something else that needs fixing here? thanks! 😄

glemaitre commented:
Mainly, I would rename index by score to be similar to other metric. Other index follow this terminology: 
https://22132-843222-gh.circle-artifacts.com/0/doc/modules/clustering.html#calinski-harabaz-index

@jnothman Do you think that this is right to do so.

jnothman commented:
Yes, it should be _score for consistency.

On 14 May 2018 at 22:50, Guillaume Lemaitre <notifications@github.com>
wrote:

> Mainly, I would rename index by score to be similar to other metric. Other
> index follow this terminology:
> https://22132-843222-gh.circle-artifacts.com/0/doc/
> modules/clustering.html#calinski-harabaz-index
>
> @jnothman <https://github.com/jnothman> Do you think that this is right
> to do so.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10827#issuecomment-388805059>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-pSlcoETdqGtVwSAxPtRNfMXTudks5tyX2VgaJpZM4Su-1P>
> .
>

logc committed:
Rename `index` to `score`

logc commented:
Rebased on top of the current `master`

logc commented:
The failing check does not seem related to these changes, but I don't know how to deal with the error:
```
60    Complete output from command python setup.py egg_info:
61    This backport is meant only for Python 2.
62    It does not work on Python 3, and Python 3 users do not need it as the concurrent.futures package is available in the standard library.
63    For projects that work on both Python 2 and 3, the dependency needs to be conditional on the Python version, like so:
64    extras_require={':python_version == "2.7"': ['futures']}
65    
66    ----------------------------------------
67Command "python setup.py egg_info" failed with error code 1 in C:\Users\appveyor\AppData\Local\Temp\1\pip-install-dzi69ulw\futures\
```

`futures` is a transitive dependency pulled in from `wheelhouse_uploader`. Any ideas?

glemaitre commented:
Weird failing. It should not be related.

logc committed:
Documentation changes

logc commented on doc/modules/clustering.rst:
```diff
@@ -1591,6 +1591,89 @@ Drawbacks
+.. math::
+   R_{ij} = \frac{s_i + s_j}{d_{ij}}
+
+Then the DB index is defined as:
```
Changed to "Davies-Bouldin"

logc commented on doc/modules/clustering.rst:
```diff
@@ -1591,6 +1591,89 @@ Drawbacks
+Zero is the lowest possible score. Values closer to zero indicate a better
+partition.
+
+In normal usage, the Davies-Bouldin index is applied to the results of a
```
I am sorry, I do not understand what is requested here (?)

logc replied:
Removed

logc replied:
I guess we can leave it explicitly as "Davies-Bouldin". "DB" might be confused with database, or DBSCAN.

glemaitre commented:
> About the what's new entry, I added one in commit cd52612 , for release 0.20 ...

Sorry I missed the file.

glemaitre commented:
I'll merge when it will be (almost) green

logc commented:
@glemaitre the checks passed this time 😄 can we merge then?

glemaitre commented:
Thanks for the recall
